src/target/riscv/riscv.c

```diff
@@ -3076,15 +3093,29 @@ static int riscv_read_memory(struct target *target, target_addr_t address,
 		return ERROR_OK;
 	}
 
-	target_addr_t physical_addr;
-	int result = target->type->virt2phys(target, address, &physical_addr);
-	if (result != ERROR_OK) {
-		LOG_TARGET_ERROR(target, "Address translation failed.");
+	int result = check_access_request(target, address, size, count);
+	if (result != ERROR_OK)
 		return result;
-	}
 
 	RISCV_INFO(r);
-	return r->read_memory(target, physical_addr, size, count, buffer, size);
+	uint32_t current_count = 0;
+	while (current_count < count) {
+		target_addr_t physical_addr;
+		result = target->type->virt2phys(target, address, &physical_addr);
+		if (result != ERROR_OK) {
+			LOG_TARGET_ERROR(target, "Address translation failed.");
+			return result;
+		}
+
+		uint32_t chunk_count = MIN(count - current_count, (RISCV_PGSIZE - RISCV_PGOFFSET(address)) / size);
```

Comment:
When mmu is not enabled, access does not need to consider crossing page boundaries.
Is it worth changing it to
```
if (mmu_enable)
  chunk_count = MIN(count - current_count, (RISCV_PGSIZE - RISCV_PGOFFSET(address)) / size);
else
  chunk_count = count;
```

Reply:
Oh, thinking about this made me realize that there is a problem with misaligned access support here. 
`( RISCV_PGSIZE - RISCV_PGOFFSET(address))/size` sets `chunk_count` to 0, so the while loop will never end.

Reply:
https://github.com/riscv-collab/riscv-openocd/pull/1149/commits/cd29d6a7219a5ac62cbed46f14762d378664c1d7